hack/generator/pkg/codegen/pipeline_remove_type_aliases.go

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"k8s.io/klog/v2"
+)
+
+// removeTypeAliases creates a pipeline stage removing type aliases
+func removeTypeAliases() PipelineStage {
+	return PipelineStage{
+		"Remove type aliases",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			visitor := astmodel.MakeTypeVisitor()
+			var result = make(astmodel.Types)
+
+			visitor.VisitTypeName = func(this *astmodel.TypeVisitor, it astmodel.TypeName, ctx interface{}) astmodel.Type {
+				resolvedName := resolveTypeName(it, types)
+				return resolvedName
+			}
+
+			for _, typeDef := range types {
+				newType := visitor.Visit(typeDef.Type(), nil)
+				result.Add(typeDef.WithType(newType))
+			}
+
+			return result, nil
+		}}
+}
+
+func resolveTypeName(name astmodel.TypeName, types astmodel.Types) astmodel.Type {
+	def, ok := types[name]
+	if !ok {
+		panic(fmt.Sprintf("Couldn't find type for type name %s", name))
+	}
+
+	// If this typeName definition has a type of object, enum, or resource
+	// it's okay. Everything else we want to pull up one level to remove the alias
+	switch concreteType := def.Type().(type) {
+	case *astmodel.ObjectType:
+		return def.Name()
+	case *astmodel.EnumType:
+		return def.Name()
+	case *astmodel.ResourceType:
+		return def.Name()
+	case astmodel.TypeName:
+		// We need to resolve further because this type is an alias
+		klog.V(3).Infof("Found type alias %s, replacing it with %s", name, concreteType)
+		return resolveTypeName(concreteType, types)
+	case *astmodel.PrimitiveType:
+		return concreteType
+	case *astmodel.OptionalType:
```

Comment:
I think for Optional, Array, and Map you want to Visit the concrete types here, or else nested TypeNames inside them will not be resolved. (This should be easy enough to write a golden test for.)

Reply:
Doesn't the default TypeVisitor also have this problem?

From what I can tell of the usage of TypeVisitor, it's usually called once for each def, and the default impl stops at a `TypeName` (it doesn't recurse in), which I think makes sense because anyway `Visit` will be called for the other types in the `Types` collection since the visitor is used in a loop?

I think this means we won't drop any types?

Reply:
but here you are inlining named types, so essentially cloning them from the collection. Depending on the processing order of the `Types` collection they may or may not have already been visited, so in some cases you will get unprocessed types inlined.

Reply:
but getting an unprocessed type inlined doesn't matter does it? That type will get processed afterwards anyway since bypassing it here won't remove it from the `Types` collection? As a matter of fact this pipeline stage shouldn't be removing any types (`len(types) == length(result)`). The types that aren't referenced anymore (because they're just a type alias and so were bypassed by this pipeline phase) will be removed by the later phase that removes unreferenced types?

Reply:
I think this calls for a unit test to prove (or disprove) the need to recurse within Optional, Array, and Map.

Reply:
George wrote one and showed me what the issue was